Ability to fetch metadata

Hacksore opened the pull request:
This will allow you to grab the metadata about a shortcut.

Commit:
Add bplist-parser dependency to this lib

Commit:
Add getMetaData method to the Shortcut class

This method will allow you to get the binary plist data converted into JSON. It will fetch the data from the downloadURL using got and then use bplist-parser to convert it to JSON.

Commit:
Exclude test.js file

I was using this file to test the implementation and don't feel it needs to make it into the repo.

Commit:
Fix JSdoc

Hacksore commented:
Working on the lint errors will push changes in a bit.

Commit:
Fix eslint errors

haykam821 commented on index.js:
```diff
@@ -73,6 +74,25 @@ class Shortcut {
+
+	/**
+	 * Gets the shortcuts metadata from the downloadURL
+	 * @returns {Object} An object with all the metadata about the shortcut
```
Do you think we could make a `ShortcutMetadata` class with a custom, more JSON-friendly format?

Hacksore replied:
Good idea!

Hacksore replied:
Not sure if the implementation I added in the latest commit makes sense. Since this is not typescript I'm unsure of how to do this correctly 😁.

Please feel free to add some feedback.

haykam821 commented on index.js:
```diff
@@ -73,6 +74,25 @@ class Shortcut {
+	}
+
+	/**
+	 * Gets the shortcuts metadata from the downloadURL
```
Punctuate this for consistency with the other documentation.

haykam821 commented on index.js:
```diff
@@ -73,6 +74,25 @@ class Shortcut {
+	 * Gets the shortcuts metadata from the downloadURL
+	 * @returns {Object} An object with all the metadata about the shortcut
+	 */
+	getMetaData() {
```
Metadata is one word, but your camel-casing puts it as meta data. Can you fix that?

Commit:
PR #3 changes

Add new ShortcutMetadata class
Fix punctuation on comment
Rename getMetaData -> getMetadata

haykam821 commented on index.js:
```diff
@@ -14,6 +15,21 @@ const baseURL = "https://www.icloud.com/shortcuts/api/records/";
+ * A shortcut metadata.
+ */
+class ShortcutMetadata {
+	constructor(object) {
```
Change `object` to `metadata` to keep it similar to the `Shortcut` constructor.

haykam821 commented on index.js:
```diff
@@ -14,6 +15,21 @@ const baseURL = "https://www.icloud.com/shortcuts/api/records/";
+ */
+class ShortcutMetadata {
+	constructor(object) {
+		// Not sure if we keep the native Apple names for the properties?
```
For these property names, I'd take off the `WFWorkflow` part and make it camel-cased, e.g. `WFWorkflowImportQuestions` would become `this.importQuestions`. For the client, it'd also probably be better to have `this.client.version` and `this.client.minimumVersion`.

haykam821 commented on index.js:
```diff
@@ -73,6 +89,28 @@ class Shortcut {
+			got(this.downloadURL, {
+				encoding: null,
+			}).then(response => {
+
```
Can you remove the two lines of whitespace between the braces and content of this block?

haykam821 commented on index.js:
```diff
@@ -91,6 +129,7 @@ class Shortcut {
  */
 function getShortcutDetails(id) {
 	return new Promise((resolve, reject) => {
+
```
Can you get rid of this extra line?

Commit:
Implement changes from #3

Fix formatting
Change ShortcutMetadata properties to have a better naming convention

Commit:
Fix eslint error

The keys for ShortcutMetadata.icon were sorted

haykam821 commented on index.js:
```diff
@@ -19,14 +19,19 @@ const baseLink = "https://www.icloud.com/shortcuts/";
-		this.WFWorkflowImportQuestions = object.WFWorkflowImportQuestions;
-		this.WFWorkflowTypes = object.WFWorkflowTypes;
-		this.WFWorkflowInputContentItemClasses = object.WFWorkflowInputContentItemClasses;
+	constructor(metadata) {
```
Can you document all these properties?

haykam821 replied:
I don't see actions in here. Can you add those, with a structure like below?

```json
{
    "actions": [
        {
            "identifier": "is.workflow.actions.date",
            "parameters": {
                "UUID": "FB44681C-9A30-43AC-A2F4-0B92272B5222"
            }
        }
    ]
}
```

Maybe make a `Action` class, and `<ShortcutMetadata>.actions` would be an array of them, with `identifier` being `WFWorkflowActionIdentifier` and `parameters` being `WFWorkflowActionParameters`. I don't think we should rename parameters, for an obvious reason (third-party apps).

haykam821 commented on index.js:
```diff
@@ -19,14 +19,19 @@ const baseLink = "https://www.icloud.com/shortcuts/";
+			minimumVersion: metadata.WFWorkflowMinimumClientVersion,
+			version: metadata.WFWorkflowClientRelease,
+		};
+		this.icon = {
```
Can these properties be named similarly to the ones in a `Shortcut`, such as `color`, `data`, and `glyph`?

haykam821 commented on index.js:
```diff
@@ -19,14 +19,19 @@ const baseLink = "https://www.icloud.com/shortcuts/";
+			imageData: metadata.WFWorkflowIcon.WFWorkflowIconImageData,
+			glyphNumber: metadata.WFWorkflowIcon.WFWorkflowIconGlyphNumber,
+		};
+		this.importQuestions = metadata.WFWorkflowImportQuestions;
```
I haven't seen the structure for them myself, but can you make a `ImportQuestion` class, so we could have an array of those in here?

Commit:
Update for PR #3

Commit:
Fix lint errors and JSdocs

Commit:
Add JSDoc properties to ImportQuestion

Commit:
Add eslint ignore file

haykam821 committed:
Make <ShortcutMetadata>.client.minimumVersion nullable

Commit:
Fix eslint error

haykam821 commented on index.js:
```diff
@@ -14,6 +15,120 @@ const baseURL = "https://www.icloud.com/shortcuts/api/records/";
+		this.category = question.Category;
+
+		/**
+		 * The action index of the qeuestion.
```
Here is a typo in the JSDoc.

haykam821 commented on index.js:
```diff
@@ -14,6 +15,120 @@ const baseURL = "https://www.icloud.com/shortcuts/api/records/";
+		this.identifier = action.WFWorkflowActionIdentifier;
+
+		/**
+		 * @type {Object[]}
```
Can you document this and specifically how we don't rename its properties?

haykam821 committed:
Delete package-lock.json

haykam821 commented on index.js:
```diff
@@ -14,6 +15,120 @@ const baseURL = "https://www.icloud.com/shortcuts/api/records/";
+
+		/**
+		 * An unknown property.
+		 * @type {array}
```
This is an array of strings.

haykam821 reviewed (changes requested):
There are only a few JSDoc-related issues, and then this pull request can be merged.

Commit:
Fix typo and JSDocs

Commit:
Merge branch 'master' of github.com:Hacksore/Shortcuts.js

haykam821 commented on index.js:
```diff
@@ -22,11 +22,15 @@ const baseLink = "https://www.icloud.com/shortcuts/";
 
 		/**
+		 * An object denoting a shortcut action.
+		 * The propeties in the object are not renamed and follow the appple naming convention.
```
There is are two typos in this property's description.

haykam821 replied:
You have fixed the second typo in 45d1f0cb7f6467f0041b3b1b22afa09d1a2c566e, but it is still lowercase when it is a company name. Also, the first typo is still there.

haykam821 commented on index.js:
```diff
@@ -22,11 +22,15 @@ const baseLink = "https://www.icloud.com/shortcuts/";
 		/**
+		 * An object denoting a shortcut action.
+		 * The propeties in the object are not renamed and follow the appple naming convention.
+		 * Example: {"WFCommentActionText":"example value"}
```
I do not think this example is necessary.

Hacksore replied:
Ok

Commit:
Fix JSDocs

Commit:
Fix typos

haykam821 committed:
Rewrite description for <Shortcut>.getMetadata

haykam821 committed:
Export new classes

haykam821 committed:
Fix ESLint key sorting error

Hacksore commented:
@haykam821 is the plan to merge #2 then come and merge #3 in?

haykam821 commented:
@Hacksore Yes, but that will mean we will need to write proper tests for the code added by this pull request.

Hacksore commented:
@haykam821 As soon as you can wrap up #2 I can spend some cycles getting tests built.

Commit:
Merge branch 'master' into master
